zipkin-storage/cassandra-v1/src/main/java/zipkin2/storage/cassandra/v1/TraceIdIndexer.java

```diff
@@ -0,0 +1,252 @@
+    final Factory factory;
+    final long expiration;
+
+    Expiration(Factory factory, K key, V value, long expiration) {
+      super(key, value);
+      this.factory = factory;
+      this.expiration = expiration;
+    }
+
+    @Override public long getDelay(TimeUnit unit) {
+      return unit.convert(expiration - factory.nanoTime(), TimeUnit.NANOSECONDS);
+    }
+
+    @Override public int compareTo(Delayed o) {
+      return Long.signum(expiration - ((Expiration) o).expiration);
+    }
+  }
+
+  final class RealTraceIdIndexer implements TraceIdIndexer {
+    static final Logger LOG = LoggerFactory.getLogger(RealTraceIdIndexer.class);
+
+    final Factory factory;
+    final Set<Input> inputs = new LinkedHashSet<>();
+
+    RealTraceIdIndexer(Factory factory) {
+      this.factory = factory;
+    }
+
+    @Override public void add(Input input) {
+      inputs.add(input);
+    }
+
+    /**
+     * The input may include inserts that already occurred, or are redundant as they don't impact
+     * QueryRequest.endTs or QueryRequest.loopback. For example, a parsed timestamp could be between
+     * timestamps of rows that already exist for a particular trace. Optimized results will be
+     * smaller when the input includes traces with local spans, or when other threads indexed the
+     * same trace.
+     */
+    @Override public Iterator<Input> iterator() {
+      Set<Input> result = entriesThatIncreaseGap();
+      if (LOG.isDebugEnabled() && inputs.size() > result.size()) {
+        int delta = inputs.size() - result.size();
+        LOG.debug("optimized out {}/{} inserts into {}", delta, inputs.size(), factory.table);
+      }
+      return result.iterator();
+    }
+
+    Set<Input> entriesThatIncreaseGap() {
+      if (inputs.isEmpty()) return inputs;
+
+      OnChangeUpdateMap toUpdate = new OnChangeUpdateMap(factory);
+      Map<Entry<String, Long>, Set<Long>> mappedInputs = new LinkedHashMap<>();
+
+      // Enter a loop that affects shared state when an update widens the time interval for a key.
+      for (Input input : inputs) {
+        Entry<String, Long> key = toEntry(input);
+        long timestamp = input.ts();
+        add(mappedInputs, key, timestamp);
+        toUpdate.currentTimestamp = timestamp;
+        factory.cache.compute(key, toUpdate);
+      }
+
+      // When the loop completes, we'll know one of our updates widened the interval of a trace, if
+      // it is the first or last timestamp. By ignoring those between an existing interval, we can
+      // end up with less Cassandra writes.
+      Set<Input> result = new LinkedHashSet<>();
+      for (Entry<String, Long> needsUpdate : toUpdate.keySet()) {
+        Expiration<Entry<String, Long>, Pair> existing = factory.cache.get(needsUpdate);
+
+        // unexpected, but possible. Something could have purged all our cached entries
+        Pair range = existing != null ? existing.getValue() : toUpdate.get(needsUpdate);
```

Comment:
Do we need to check if `existing` is expired here?

Reply:
well since we just wrote it, it couldn't be purged due to expiry (expiration unit is seconds). the only thing possible is something else swamped the cache so it is not here due to the bounds of it (which is likely a tuning problem I suppose).

If it is expired we don't know if it is redundant or not is the main thing, so lack of knowledge means we should write it. I'll try to clarify in the comments.